### PR TITLE
Introduce plugin for medline (.nbib) format

### LINKED ISSUE #21: Support of PubMed format
PubMed has a new interface and  is still keeping the legacy PubMed which may be discontinued at any time. We need to make sure that our users can get data from PubMed to Rayyan. 
PubMed created their own format in their new interface, called PubMed (has some similarities with the RIS format)!
For reference, see the below communication with US NLM support

Hi Dr. Ouzzani, 

Thank you for writing to the help desk.  We do not anticipate adding XML to the new PubMed web interface. PubMed data is available in XML format through our E-utilities APIs (https://www.ncbi.nlm.nih.gov/books/NBK25497/) and our Data Distribution program (https://pubmed.ncbi.nlm.nih.gov/help/#download-pubmed-data).

An example EFetch URL that retrieves XML data for 2 PMIDs: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=11748933,11700088&retmode=xml

A new API will be developed for PubMed; however, we do not yet have a timeline to share regarding the update. Please subscribe to the E-Utilities Announcement Mailing List: https://www.ncbi.nlm.nih.gov/mailman/listinfo/utilities-announce/

From the Save menu, you can also use the "PubMed" format which has replaced RIS.  This file will include the abstracts.

Kind regards, 
Marie Collins
U.S. National Library of Medicine
****
* PLEASE DO NOT MODIFY THE SUBJECT LINE OF THIS EMAIL WHEN RESPONDING TO ENSURE CORRECT TRACKING *

Case Information:
Case #: CAS-514295-C2Q7Z7
Customer Name: Mourad OUZZANI
Customer Email: mouzzani@hbku.edu.qa
Case Created: 2020-04-01T12:49:27Z

Summary: Saving results

Details: Hi,

In the legacy PubMed, one can save results using your XML format which was handy for getting all the bibliographic information including asbtracts and using them in other systems (such as on rayyan.qcri.org) then in the new PubMED I've used the RIS format few times but now it disappeared!
It will be great to have a standard reference format that would include abstracts such as RIS or CIW or simply the CSV format. The previous XML format would work too.

Best,
Mourad Ouzzani, PhD

### PR DESCRIPTION
**Pull request by fabubaker**
Adds plugin for the PubMed MEDLINE format outlined [here](https://www.nlm.nih.gov/bsd/mms/medlineelements.html#lid).

Fixes #21.

### CODE REVIEW

File: lib/rayyan-formats-plugins/plugins/medline.rb [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+module RayyanFormats
+  module Plugins
+    class MEDLINE < RayyanFormats::Base
+      title 'MEDLINE'
+      extension 'nbib'
+      description 'Supports the Pubmed MEDLINE format as outlined here: https://www.nlm.nih.gov/bsd/mms/medlineelements.html'
+
+      detect do |first_line, _|
+        first_line.start_with?('PMID')
```
[the remark this answers is not in the document]
Comment (fabubaker) [reply lib/rayyan-formats-plugins/plugins/medline.rb:11]: It says it is present on all records but I'm not sure if records have to start with it. This is purely from downloading samples from PubMed, all of them start with PMID.

What should I do here?

File: spec/plugins/medline_spec.rb [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+          expect(target.date_array).to eq(%w[2017 2])
+          expect(target.journal_title).to eq('Medizinische Monatsschrift fur Pharmazeuten')
+          expect(target.journal_issn).to eq('0342-9601 (Print)')
+          expect(target.journal_abbreviation).to eq('Med Monatsschr Pharm')
+          expect(target.jvolume).to eq(40)
+          expect(target.jissue).to eq(2)
+          expect(target.pagination).to eq('55-64')
+          expect(target.authors).to eq(['Wörmann B'])
+          expect(target.affiliation).to eq(['Medizinische Klinik mit Schwerpunkt Hämatologie'])
+          expect(target.url).to eq(MEDLINE.send(:get_pubmed_url, '29952495'))
```
[the remark this answers is not in the document]
Comment (fabubaker) [reply spec/plugins/medline_spec.rb:41]: I was trying to be smart about it and not have the test case depend on a hard-coded url.

File: lib/rayyan-formats-plugins/plugins/medline.rb [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+          target.date_array = get_date_array article['DP']
+          target.journal_title = article['JT']
+          target.journal_issn = get_issn article['IS']
+          target.journal_abbreviation = article['TA']
+          # to_i grabs the first integer it sees, so '5 Spec No' becomes 5
+          target.jvolume = article['VI']&.to_i
+          target.jissue = article['IP']&.to_i
+          target.pagination = article['PG']
+          target.authors = Array(article['AU'])
+          target.affiliation = get_multivalued_field_with_newline_merger article['AD']
```
[the remark this answers is not in the document]
Comment (fabubaker) [reply lib/rayyan-formats-plugins/plugins/medline.rb:34]: You want to truncate the abstracts line when exporting? I guess I can make an inverse function but other formats also pack an abstract into one long line so I think it should be fine no?

File: lib/rayyan-formats-plugins/plugins/medline.rb [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        end
+
+        def get_issn(issn)
+          # MEDLINE returns multiple ISSN, return the first one
+          return unless issn
+
+          issn.is_a?(Array) ? issn[0] : issn
+        end
+
+        def get_multivalued_field_with_newline_merger(abstracts)
```
[the remark this answers is not in the document]
Comment (fabubaker) [reply lib/rayyan-formats-plugins/plugins/medline.rb]: Fixed.

File: lib/rayyan-formats-plugins/plugins/medline.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (hammady) [reply lib/rayyan-formats-plugins/plugins/medline.rb:11]: We will assume it should be there. I tried BibDesk and it refuses to import the file if it doesn't start with `PMID`. It also ignores all entries that do not start with `PMID`.

**Comment by mouradvt04** [comment]
ERIC (https://eric.ed.gov/) allows saving search results to the PubMed format (Create a file for use with citation management software, in a MEDLINE/PubMed-style (.nbib) format.) but I just tried a file from ERIC and it wasn't recognized by Rayyan. One difference I see is the absence of the tag "PMID" which I guess is normal.  Here is the file from ERIC:

[ERIC2020-07-22_02.45.22.txt](https://github.com/rayyanqcri/rayyan-formats-plugins/files/4958347/ERIC2020-07-22_02.45.22.txt)

**Comment by fabubaker** [comment]
The PMID tag is what we use to detect whether the format is PubMed or not. I guess we can parse the `OWN - ERIC` line  to support the ERIC format.

**Comment by mouradvt04** [comment]
Why not "IF PMID ELIF OWN"?
I think it would be sensible to do it as other databases may adapt this format and won't have PMID but would have OWN.

**Comment by mouradvt04** [comment]
Another case is when using PubMed Central PMC (https://www.ncbi.nlm.nih.gov/pmc/), searches can be save in the MEDLINE format (which is the same as the PubMed format we're supporting here). The file start with something like  "PMC - PMC7370876"  (PMC ID) and in many cases it will have a  second line such as "PMID- 32690910". According to https://www.nlm.nih.gov/bsd/difference.html "Some PMC content, such as book reviews, is not cited in PubMed."
So if we support this case "IF PMC" then we're also covering PMC.

**Pull request merged**